test/e2e/storage/host_path_type.go

```diff
@@ -0,0 +1,332 @@
+		})
+
+		It("Should fail on mounting character device 'achardev' when HostPathType is HostPathBlockDev", func() {
+			verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, targetCharDev, fmt.Sprintf("%s is not a block device", targetCharDev), &hostPathBlockDev)
+		})
+	})
+
+	Context("HostPathType Block Device Test", func() {
+		var targetBlockDev string
+
+		BeforeEach(func() {
+			targetBlockDev = path.Join(hostBaseDir, "ablkdev")
+
+			By("Create a block device for further testing")
+			_, err := utils.PodExec(basePod, fmt.Sprintf("mknod %s b 89 1", path.Join(mountBaseDir, "ablkdev")))
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("Should fail on mounting non-existent block device 'does-not-exist-blk-dev' when HostPathType is HostPathBlockDev", func() {
+			blkDevPath := path.Join(hostBaseDir, "does-not-exist-blk-dev")
+			verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, blkDevPath, fmt.Sprintf("%s is not a block device", blkDevPath), &hostPathBlockDev)
+		})
+
+		It("Should be able to mount block device 'ablkdev' successfully when HostPathType is HostPathBlockDev", func() {
+			verifyPodHostPathType(f, basePod.Spec.NodeName, targetBlockDev, &hostPathBlockDev)
+		})
+
+		It("Should be able to mount block device 'ablkdev' successfully when HostPathType is HostPathUnset", func() {
+			verifyPodHostPathType(f, basePod.Spec.NodeName, targetBlockDev, &hostPathUnset)
+		})
+
+		It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathDirectory", func() {
+			verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, targetBlockDev, fmt.Sprintf("%s is not a directory", targetBlockDev), &hostPathDirectory)
+		})
+
+		It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathFile", func() {
+			verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, targetBlockDev, fmt.Sprintf("%s is not a file", targetBlockDev), &hostPathFile)
+		})
+
+		It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathSocket", func() {
+			verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, targetBlockDev, fmt.Sprintf("%s is not a socket", targetBlockDev), &hostPathSocket)
+		})
+
+		It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathCharDev", func() {
+			verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, targetBlockDev, fmt.Sprintf("%s is not a character device", targetBlockDev), &hostPathCharDev)
+		})
+	})
+})
+
+func newHostPathTypeTestPod(nodeName, hostDir, mountDir string, hostPathType *v1.HostPathType) *v1.Pod {
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "test-hostpath-type-",
+		},
+		Spec: v1.PodSpec{
+			NodeName:      nodeName,
+			RestartPolicy: v1.RestartPolicyNever,
+			Containers: []v1.Container{
+				{
+					Name:    "host-path-testing",
+					Image:   imageutils.GetE2EImage(imageutils.Hostexec),
```

Comment:
I'm not familiar with this hostexec image, but in the past, we've seen issues with busybox images ignoring sigterm signals.  Tests using them would take a long time to delete the pod because it has to wait for the podTerminationGracePeriod.  We fixed this by using an image that behaves better, like nginx.

Reply:
@msau42 A socket file needs to be created by issuing command `nc -lU` in this container image. For nginx and busybox, flag `-U` is not supported.

I've tried several images and `hostexec` is the chosen one.

[`hostexec`](https://github.com/kubernetes/kubernetes/blob/master/test/images/hostexec/Dockerfile) is a small image based on alpine.

Reply:
I remember @cofyc ran into a similar issue and added a trap for sigterm to make the pod shutdown faster. Maybe something similar can be done here

Reply:
FYI https://github.com/kubernetes/kubernetes/pull/64015, in this PR I explained why some container commands do not exit immediately on SIGTERM (docker stop), it's better to register a exit handler for SIGTERM if process didn't do it.

```
trap exit TERM
```